rmw_connext_dynamic_cpp/src/functions.cpp

```diff
@@ -56,17 +59,45 @@
 #include "rosidl_typesupport_introspection_cpp/service_introspection.hpp"
 #include "rosidl_typesupport_introspection_cpp/visibility_control.h"
 
+#include "rosidl_typesupport_introspection_c/field_types.h"
+#include "rosidl_typesupport_introspection_c/identifier.h"
+#include "rosidl_typesupport_introspection_c/message_introspection.h"
+#include "rosidl_typesupport_introspection_c/service_introspection.h"
+#include "rosidl_typesupport_introspection_c/visibility_control.h"
+
 #include "rmw_connext_shared_cpp/shared_functions.hpp"
 #include "rmw_connext_shared_cpp/types.hpp"
 
+bool using_introspection_c_typesupport(const char * typesupport_identifier)
+{
+  return strcmp(typesupport_identifier, rosidl_typesupport_introspection_c__identifier) == 0;
+}
+
+bool using_introspection_cpp_typesupport(const char * typesupport_identifier)
+{
+  return strcmp(
+    typesupport_identifier,
+    rosidl_typesupport_introspection_cpp::typesupport_introspection_identifier) == 0;
+}
+
 ROSIDL_TYPESUPPORT_INTROSPECTION_CPP_LOCAL
 inline std::string
 _create_type_name(
-  const rosidl_typesupport_introspection_cpp::MessageMembers * members,
-  const std::string & sep)
+  const void * untyped_members,
+  const std::string & sep,
+  const char * typesupport)
 {
+  std::string package_name, message_name;
+  if (using_introspection_c_typesupport(typesupport)) {
+    CREATE_TYPENAME_PREFIX(INTROSPECTION_C_TYPE)
+  } else if (using_introspection_cpp_typesupport(typesupport)) {
+    CREATE_TYPENAME_PREFIX(INTROSPECTION_CPP_TYPE)
```

Comment:
Since both macros have the code to access the fields of `members` in common I would suggest to let the macro only define the `members` variable and continue to handle accessing the fields `package_name_` and `message_name_` outside.